### PR TITLE
Glibc update to support M98

### LINKED ISSUE #6643: Glibc Throws error upon attempt to download
## Description
Trying to install glibc returns an error.

If you executed a specific command:
```shell
crew install glibc
```
or 
```shell
crew install -s glibc
```

Output:
```shell
glibc: The GNU C Library project provides the core libraries for GNU/Linux systems.
https://www.gnu.org/software/libc/
Version: 
License: LGPL-2.1+, BSD, HPND, ISC, inner-net, rc, and PCRE
Performing pre-flight checks...
/usr/local/lib64/ruby/3.0.0/uri/rfc3986_parser.rb:18:in `rescue in split': bad URI(is not URI?): nil (URI::InvalidURIError)
        from /usr/local/lib64/ruby/3.0.0/uri/rfc3986_parser.rb:15:in `split'
        from /usr/local/lib64/ruby/3.0.0/uri/rfc3986_parser.rb:72:in `parse'
        from /usr/local/lib64/ruby/3.0.0/uri/common.rb:171:in `parse'
        from /usr/local/bin/crew:638:in `download'
        from /usr/local/bin/crew:1315:in `install'
        from /usr/local/bin/crew:1181:in `resolve_dependencies_and_install'
        from /usr/local/bin/crew:1602:in `block in install_command'
        from /usr/local/bin/crew:1597:in `each'
        from /usr/local/bin/crew:1597:in `install_command'
        from /usr/local/bin/crew:1690:in `<main>'
/usr/local/lib64/ruby/3.0.0/uri/rfc3986_parser.rb:16:in `split': undefined method `to_str' for nil:NilClass (NoMethodError)
        from /usr/local/lib64/ruby/3.0.0/uri/rfc3986_parser.rb:72:in `parse'
        from /usr/local/lib64/ruby/3.0.0/uri/common.rb:171:in `parse'
        from /usr/local/bin/crew:638:in `download'
        from /usr/local/bin/crew:1315:in `install'
        from /usr/local/bin/crew:1181:in `resolve_dependencies_and_install'
        from /usr/local/bin/crew:1602:in `block in install_command'
        from /usr/local/bin/crew:1597:in `each'
        from /usr/local/bin/crew:1597:in `install_command'
        from /usr/local/bin/crew:1690:in `<main>'
```

## Versions and system information
chromebrew: `1.21.8`

ChromeOS: `cat /etc/lsb-release`
```
CHROMEOS_ARC_ANDROID_SDK_VERSION=28
CHROMEOS_ARC_VERSION=8050614
CHROMEOS_AUSERVER=https://tools.google.com/service/update2
CHROMEOS_BOARD_APPID={E9E5C8F0-C30C-190F-CA75-D6184C3FEED6}
CHROMEOS_CANARY_APPID={90F229CE-83E2-4FAF-8479-E368A34938B1}
CHROMEOS_DEVSERVER=
CHROMEOS_RELEASE_APPID={E9E5C8F0-C30C-190F-CA75-D6184C3FEED6}
CHROMEOS_RELEASE_BOARD=setzer-signed-mpkeys
CHROMEOS_RELEASE_BRANCH_NUMBER=27
CHROMEOS_RELEASE_BUILDER_PATH=setzer-release/R98-14388.27.0
CHROMEOS_RELEASE_BUILD_NUMBER=14388
CHROMEOS_RELEASE_BUILD_TYPE=Official Build
CHROMEOS_RELEASE_CHROME_MILESTONE=98
CHROMEOS_RELEASE_DESCRIPTION=14388.27.0 (Official Build) beta-channel setzer 
CHROMEOS_RELEASE_KEYSET=mp
CHROMEOS_RELEASE_NAME=Chrome OS
CHROMEOS_RELEASE_PATCH_NUMBER=0
CHROMEOS_RELEASE_TRACK=beta-channel
CHROMEOS_RELEASE_VERSION=14388.27.0
DEVICETYPE=CHROMEBOOK
GOOGLE_RELEASE=14388.27.0
```
Model: HP Chromebook
Architecture: `x86_64`

### PR DESCRIPTION
**Pull request by satmandu**
Fixes #6643 

- Glibc 2.33 appears to work for people.
- glibc 2.27 builds keep segfaulting, so reverted to older binaries.

Works properly:
- [x] x86_64
- [x] armv7l
- [x] i686

### Run the following to get this pull request's changes locally for testing.
```
CREW_TESTING_REPO=https://github.com/satmandu/chromebrew.git CREW_TESTING_BRANCH=glibc_test CREW_TESTING=1 crew update
```
If your system is broken because you just updated to M98 and you NEED glibc 2.33, try this:
Do this if you are on `x86_64`.
```
LD_LIBRARY_PATH=/usr/lib64 /usr/bin/curl -Ls https://gitlab.com/api/v4/projects/26210301/packages/generic/glibc/2.33_x86_64/glibc-2.33-chromeos-x86_64.tpxz -o /tmp/glibc-2.33-chromeos-x86_64.tpxz
LD_LIBRARY_PATH=/usr/lib64 tar -Ipixz -xf /tmp/glibc-2.33-chromeos-x86_64.tpxz --exclude=dlist --exclude=filelist -C /
CREW_TESTING_REPO=https://github.com/satmandu/chromebrew.git CREW_TESTING_BRANCH=glibc_test CREW_TESTING=1 crew update
crew upgrade
```
And if on `armv7l` do this.
```
LD_LIBRARY_PATH=/usr/lib /usr/bin/curl -Ls https://gitlab.com/api/v4/projects/26210301/packages/generic/glibc/2.33_armv7l/glibc-2.33-chromeos-armv7l.tpxz -o /tmp/glibc-2.33-chromeos-armv7l.tpxz
LD_LIBRARY_PATH=/usr/lib tar -Ipixz -xf /tmp/glibc-2.33-chromeos-armv7l.tpxz --exclude=dlist --exclude=filelist -C /
CREW_TESTING_REPO=https://github.com/satmandu/chromebrew.git CREW_TESTING_BRANCH=glibc_test CREW_TESTING=1 crew update
crew upgrade
```

If installing glibc 2.33 breaks your system:
Do this if you are on `x86_64`.
```
LD_LIBRARY_PATH=/usr/lib64 /usr/bin/curl -Ls https://gitlab.com/api/v4/projects/26210301/packages/generic/glibc/2.32-2_x86_64/glibc-2.32-2-chromeos-x86_64.tpxz -o /tmp/glibc-2.32-2-chromeos-x86_64.tpxz
LD_LIBRARY_PATH=/usr/lib64 tar -Ipixz -xf /tmp/glibc-2.32-2-chromeos-x86_64.tpxz --exclude=dlist --exclude=filelist -C /
crew postinstall glibc ; crew postinstall glibc
```
And if on `armv7l` do this.
```
LD_LIBRARY_PATH=/usr/lib /usr/bin/curl -Ls https://gitlab.com/api/v4/projects/26210301/packages/generic/glibc/2.32-2_armv7l/glibc-2.32-2-chromeos-armv7l.tpxz -o /tmp/glibc-2.32-2-chromeos-armv7l.tpxz
LD_LIBRARY_PATH=/usr/lib tar -Ipixz -xf /tmp/glibc-2.32-2-chromeos-armv7l.tpxz --exclude=dlist --exclude=filelist -C /
crew postinstall glibc ; crew postinstall glibc
```

### CODE REVIEW

**Comment by uberhacker** [comment]
Can this be converted to draft if it's still a WIP?

File: packages/glibc.rb [hunk 1]
```diff
@@ -98,58 +289,109 @@ def self.patch
...
+      system 'curl -Ls https://dev.gentoo.org/~dilfridge/distfiles/glibc-2.33-patches-6.tar.xz | \
+        hashpipe sha256 29c2e4036c2b33b830a9588055c63fde5dd5255bcfd5fad2fc92f3bbd27456c1 | \
+        tar -xJf - -C gentoopatches'
+      Dir.glob('gentoopatches/patches/*.patch').each do |patch|
+        system "patch -Np1 -i #{patch} || true"
+      end
+      @googlesource_branch = 'release-R99-14469.B'
+      system "git clone --depth=1 -b  #{@googlesource_branch} https://chromium.googlesource.com/chromiumos/overlays/chromiumos-overlay googlesource"
+      Dir.glob('googlesource/sys-libs/glibc/files/local/glibc-2.33/*.patch').each do |patch|
+        puts patch
```
Comment (uberhacker) [inline packages/glibc.rb:320]: Add `if @verbose`?

File: packages/glibc.rb [hunk 2]
```diff
@@ -98,58 +289,109 @@ def self.patch
...
+      system "sed -i 's,/lib/locale,/lib#{CREW_LIB_SUFFIX}/locale,g' fedora/build-locale-archive.c"
+      system "sed -i 's,/usr/sbin/tzdata-update,/bin/true,g' fedora/build-locale-archive.c"
+      system "sed -i 's,verbose,locale_verbose,g' fedora/build-locale-archive.c"
+      system "sed -i 's,be_quiet,locale_be_quiet,g' fedora/build-locale-archive.c"
+      FileUtils.mkdir_p 'gentoopatches'
+      system 'curl -Ls https://dev.gentoo.org/~dilfridge/distfiles/glibc-2.33-patches-6.tar.xz | \
+        hashpipe sha256 29c2e4036c2b33b830a9588055c63fde5dd5255bcfd5fad2fc92f3bbd27456c1 | \
+        tar -xJf - -C gentoopatches'
+      Dir.glob('gentoopatches/patches/*.patch').each do |patch|
+        system "patch -Np1 -i #{patch} || true"
```
Comment (uberhacker) [inline packages/glibc.rb:315]: Add `puts "patch -Np1 -i #{patch} || true" if @verbose`?

File: packages/glibc.rb [hunk 3]
```diff
@@ -258,20 +503,82 @@ def self.install
...
+    # make[2]: Leaving directory '/usr/local/tmp/crew/glibc.20211222010048.dir/glibc-2.23/elf'
+    # make[1]: *** [Makefile:214: elf/tests] Error 2
+    # make[1]: Target 'check' not remade because of errors.
+    # make[1]: Leaving directory '/usr/local/tmp/crew/glibc.20211222010048.dir/glibc-2.23'
+    # make: *** [Makefile:9: check] Error 2
+    # return if ARCH == 'i686'
+
+    # Dir.chdir 'glibc_build' do
+    # system "make -k -j#{CREW_NPROC} check || true"
+    # end
```
Comment (uberhacker) [inline packages/glibc.rb:580]: I take it the tests fail?  Can we add a note to explain the error and why it was commented?

Comment (satmandu) [reply packages/glibc.rb:580]: Sure

File: packages/glibc.rb [hunk 4]
```diff
@@ -7,28 +7,30 @@ class Glibc < Package
...
-      i686: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/glibc/2.23_i686/glibc-2.23-chromeos-i686.tar.xz'
+      i686: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/glibc/2.23-3_i686/glibc-2.23-3-chromeos-i686.tar.xz'
     })
     binary_sha256({
-      i686: '52145b65cb49c2751f69d4c46636f0685f2abb6685d8a080b71b2f091595a950'
+      i686: '3ee19cbb907eb219a2c1b02df6de1ca13b09b0d375101657d54a2485aacdc445'
     })
-  when '2.27'
-    version LIBC_VERSION
+  elsif @LIBC_VERSION == '2.27'
```
Comment (uberhacker) [inline packages/glibc.rb:32]: Can constants be instance variables?  Why do we need the `@`?

File: packages/glibc.rb [hunk 5]
```diff
@@ -7,28 +7,30 @@ class Glibc < Package
...
 
-  # enable-obsolete-rpc should NOT be included as a build option as
-  # this functionality is provided by libtirpc
+  no_env_options
+  conflicts_ok
 
-  case LIBC_VERSION
-  when '2.23'
-    version LIBC_VERSION
+  @LIBC_VERSION = LIBC_VERSION
```
Comment (uberhacker) [inline packages/glibc.rb:17]: Why is this necessary?

File: packages/glibc.rb (hunk 4 above)
Comment (satmandu) [reply packages/glibc.rb:32]: This will let you set the instance variable at the top so you can do cross-glibc-version builds.

File: packages/glibc.rb (hunk 5 above)
Comment (satmandu) [reply packages/glibc.rb:17]: This will let you set the instance variable at the top so you can do cross-glibc-version builds, say you are on glibc 2.27 and want to build a glibc for 2.33, but you are not on a glibc 2.33 system.

File: packages/glibc.rb (hunk 3 above)
Comment (satmandu) [reply packages/glibc.rb]: Comments added.

File: packages/glibc.rb (hunk 4 above)
Comment (satmandu) [reply packages/glibc.rb:32]: maybe these should all be lowercase?

File: packages/glibc.rb [hunk 6]
```diff
@@ -98,58 +289,109 @@ def self.patch
...
+      system "sed -i 's,/lib/locale,/lib#{CREW_LIB_SUFFIX}/locale,g' fedora/build-locale-archive.c"
+      system "sed -i 's,/usr/sbin/tzdata-update,/bin/true,g' fedora/build-locale-archive.c"
+      system "sed -i 's,verbose,locale_verbose,g' fedora/build-locale-archive.c"
+      system "sed -i 's,be_quiet,locale_be_quiet,g' fedora/build-locale-archive.c"
+      FileUtils.mkdir_p 'gentoopatches'
+      system 'curl -Ls https://dev.gentoo.org/~dilfridge/distfiles/glibc-2.33-patches-6.tar.xz | \
+        hashpipe sha256 29c2e4036c2b33b830a9588055c63fde5dd5255bcfd5fad2fc92f3bbd27456c1 | \
+        tar -xJf - -C gentoopatches'
+      Dir.glob('gentoopatches/patches/*.patch').each do |patch|
+        system "patch -Np1 -i #{patch} || true" if @verbose
```
Comment (uberhacker) [inline packages/glibc.rb:315]: The patch is not applied unless verbose options is provided?  Shouldn't this be a `puts` statement instead?

File: packages/glibc.rb [hunk 7]
```diff
@@ -42,42 +44,231 @@ class Glibc < Package
...
+      GLIBC_223_HEREDOC
+      case ARCH
+      when 'i686'
+        File.write('glibc_223_i686.patch', @glibc_223_i686_patch)
+        system 'patch -Np1 -i glibc_223_i686.patch'
+      when 'armv7l', 'x86_64'
+        @googlesource_branch = 'release-R91-13904.B'
+        system "git clone --depth=1 -b  #{@googlesource_branch} https://chromium.googlesource.com/chromiumos/overlays/chromiumos-overlay googlesource"
+        Dir.glob("googlesource/sys-libs/glibc/files/local/glibc-2.27/glibc-#{@libc_version}*.patch").each do |patch|
+        puts patch
```
Comment (uberhacker) [inline packages/glibc.rb:265]: Change to `puts "patch -Np1 -i #{patch}" if @verbose`?

File: packages/glibc.rb (hunk 6 above)
Comment (satmandu) [reply packages/glibc.rb]: Fixed.

File: packages/glibc.rb [hunk 8]
```diff
@@ -98,58 +289,109 @@ def self.patch
...
+      system 'curl -Ls https://dev.gentoo.org/~dilfridge/distfiles/glibc-2.33-patches-6.tar.xz | \
+        hashpipe sha256 29c2e4036c2b33b830a9588055c63fde5dd5255bcfd5fad2fc92f3bbd27456c1 | \
+        tar -xJf - -C gentoopatches'
+      Dir.glob('gentoopatches/patches/*.patch').each do |patch|
+        system "patch -Np1 -i #{patch} || true" if @verbose
+      end
+      @googlesource_branch = 'release-R99-14469.B'
+      system "git clone --depth=1 -b  #{@googlesource_branch} https://chromium.googlesource.com/chromiumos/overlays/chromiumos-overlay googlesource"
+      Dir.glob('googlesource/sys-libs/glibc/files/local/glibc-2.33/*.patch').each do |patch|
+        puts patch if @verbose
```
Comment (uberhacker) [inline packages/glibc.rb]: Change to `puts "patch -Np1 < #{patch} || true" if @verbose`?

File: packages/glibc.rb [hunk 9]
```diff
@@ -96,60 +287,113 @@ def self.patch
...
-        system "patch -Np1 -i #{patch}"
+        puts "patch -Np1 < #{patch}" if verbose
+        system "patch -Np1 < #{patch}"
       end
-      @googlesource_branch = 'release-R92-13982.B'
+      @googlesource_branch = 'release-R96-14268.B'
       system "git clone --depth=1 -b  #{@googlesource_branch} https://chromium.googlesource.com/chromiumos/overlays/chromiumos-overlay googlesource"
       Dir.glob('googlesource/sys-libs/glibc/files/local/glibc-2.32/*.patch').each do |patch|
-        puts patch
+        puts "patch -Np1 < #{patch}" if verbose
```
Comment (uberhacker) [inline packages/glibc.rb:296]: `@verbose`?

File: packages/glibc.rb [hunk 10]
```diff
@@ -42,42 +44,231 @@ class Glibc < Package
...
+        +config_vars="$config_vars
+        +with-lld = $libc_cv_with_lld"
+        +
+
+         # These programs are version sensitive.
+      GLIBC_223_HEREDOC
+      case ARCH
+      when 'i686'
+        File.write('glibc_223_i686.patch', @glibc_223_i686_patch)
+        system 'patch -Np1 -i glibc_223_i686.patch'
```
Comment (uberhacker) [inline packages/glibc.rb:260]: Add `puts 'patch -Np1 -i glibc_223_i686.patch' if @verbose`?
